### PR TITLE
schemadiff: normalize table options case

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Follow up to https://github.com/vitessio/vitess/pull/10188, this PR normalizes values of table options, upon reading table definition. 

for example, the engine name for InnoDB is always `InnoDB` at that specific case. Even though `INNODB` is valid, we will always transform it to `InnoDB`.

Character sets and collations are always lower cased. `ROW_FORMAT` values are always upper cased.

We also now explicitly mutate `utf8` to `utf8mb3`.

## Related Issue(s)

- #10203 
- #10188 
- Depends on https://github.com/vitessio/vitess/pull/10197 to merge first

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schemadiff/table.go [hunk 1]
```diff
@@ -167,29 +166,110 @@ func NewCreateTableEntity(c *sqlparser.CreateTable) *CreateTableEntity {
...
+	}
+}
+
+var integralTypes = map[string]bool{
+	"tinyint":   true,
+	"smallint":  true,
+	"mediumint": true,
+	"int":       true,
+	"bigint":    true,
+}
```
Comment (shlomi-noach) [inline go/vt/schemadiff/table.go:198]: Let's move these to `mysql.go`. There's other stuff I will move there.

**Comment by shlomi-noach** [comment]
Tracking issue: https://github.com/vitessio/vitess/issues/10203

**Comment by shlomi-noach** [comment]
We have a legitimate `endtoend` test failure: https://github.com/vitessio/vitess/runs/6272304441?check_suite_focus=true

```log
2022-05-03T11:53:18.3537021Z === RUN   TestSchemaChange/datetime-to-timestamp/validate_diff
2022-05-03T11:53:18.3537463Z     schemadiff_vrepl_suite_test.go:366: 
2022-05-03T11:53:18.3537981Z         	Error Trace:	schemadiff_vrepl_suite_test.go:366
2022-05-03T11:53:18.3538717Z         	            				schemadiff_vrepl_suite_test.go:256
2022-05-03T11:53:18.3539103Z         	Error:      	Not equal: 
2022-05-03T11:53:18.3542510Z         	            	expected: "CREATE TABLE `onlineddl_test` (\n  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,\n  `i` int(11) NOT NULL,\n  `ts0` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,\n  `ts1` timestamp NULL DEFAULT NULL,\n  `dt2` datetime DEFAULT NULL,\n  `t` timestamp NULL DEFAULT NULL,\n  `updated` tinyint(3) unsigned DEFAULT '0',\n  PRIMARY KEY (`id`),\n  KEY `i_idx` (`i`)\n) ENGINE=InnoDB DEFAULT CHARSET=utf8"
2022-05-03T11:53:18.3546522Z         	            	actual  : "CREATE TABLE `onlineddl_test` (\n  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,\n  `i` int(11) NOT NULL,\n  `ts0` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,\n  `ts1` timestamp NULL DEFAULT NULL,\n  `dt2` datetime DEFAULT NULL,\n  `t` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',\n  `updated` tinyint(3) unsigned DEFAULT '0',\n  PRIMARY KEY (`id`),\n  KEY `i_idx` (`i`)\n) ENGINE=InnoDB DEFAULT CHARSET=utf8"
2022-05-03T11:53:18.3547431Z         	            	
2022-05-03T11:53:18.3547788Z         	            	Diff:
2022-05-03T11:53:18.3548274Z         	            	--- Expected
2022-05-03T11:53:18.3548654Z         	            	+++ Actual
2022-05-03T11:53:18.3549138Z         	            	@@ -6,3 +6,3 @@
2022-05-03T11:53:18.3549705Z         	            	   `dt2` datetime DEFAULT NULL,
2022-05-03T11:53:18.3550358Z         	            	-  `t` timestamp NULL DEFAULT NULL,
2022-05-03T11:53:18.3551225Z         	            	+  `t` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
2022-05-03T11:53:18.3552051Z         	            	   `updated` tinyint(3) unsigned DEFAULT '0',
2022-05-03T11:53:18.3552734Z         	Test:       	TestSchemaChange/datetime-to-timestamp/validate_diff
2022-05-03T11:53:18.3553517Z         	Messages:   	mismatched table structure. ALTER query was: ALTER TABLE `onlineddl_test` MODIFY COLUMN `t` timestamp
2022-05-03T11:53:18.3554015Z     schemadiff_vrepl_suite_test.go:376: 
2022-05-03T11:53:18.3554523Z         	Error Trace:	schemadiff_vrepl_suite_test.go:376
2022-05-03T11:53:18.3555260Z         	            				schemadiff_vrepl_suite_test.go:256
2022-05-03T11:53:18.3556472Z         	Error:      	Expected nil, but got: &schemadiff.AlterTableEntityDiff{from:(*schemadiff.CreateTableEntity)(0xc0003fef00), to:(*schemadiff.CreateTableEntity)(0xc0003ff270), alterTable:(*sqlparser.AlterTable)(0xc000476c60)}
2022-05-03T11:53:18.3557412Z         	Test:       	TestSchemaChange/datetime-to-timestamp/validate_diff
```

File: go/vt/schemadiff/mysql.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	"varchar":    true,
+	"text":       true,
+	"tinytext":   true,
+	"mediumtext": true,
+	"longtext":   true,
+	"enum":       true,
+	"set":        true,
+}
+
+// This is the default charset for MySQL 8.0
```
Comment (dbussink) [inline go/vt/schemadiff/mysql.go:47]: This means that right now `schemadiff` is opinionated in the sense that it targets specifically how MySQL 8.0 works.

File: go/vt/schemadiff/table.go [hunk 3]
```diff
@@ -163,8 +162,159 @@ func NewCreateTableEntity(c *sqlparser.CreateTable) *CreateTableEntity {
...
+		col.Type.Options.Collate = strings.ToLower(col.Type.Options.Collate)
+
+		// See https://dev.mysql.com/doc/refman/8.0/en/create-table.html
+		// If neither NULL nor NOT NULL is specified, the column is treated as though NULL had been specified.
+		// That documentation though is not 100% true. There's an exception, and that is
+		// the `explicit_defaults_for_timestamp` flag. When that is disabled (the default on 5.7),
+		// a timestamp defaults to `NOT NULL`.
+		//
+		// We opt here to instead remove that difference and always then add `NULL` and treat
+		// `explicit_defaults_for_timestamp` as always enabled in the context of DDL for diffing.
```
Comment (dbussink) [inline go/vt/schemadiff/table.go:226]: This is another case where the MySQL 8.0 behavior is considered the standard and the behavior we assume.

**Review by dbussink — APPROVED** [review]
The changes here now better normalize / simplify DDL which means few cases of different behavior based on underlying MySQL (patch) versions and fewer potential conflicts when diffing.

File: go/vt/schemadiff/mysql.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	"swe7":     "swe7_swedish_ci",
+	"tis620":   "tis620_thai_ci",
+	"ucs2":     "ucs2_general_ci",
+	"ujis":     "ujis_japanese_ci",
+	"utf16":    "utf16_general_ci",
+	"utf16le":  "utf16le_general_ci",
+	"utf32":    "utf32_general_ci",
+	"utf8mb3":  "utf8_general_ci",
+	"utf8mb4":  "utf8mb4_0900_ai_ci",
+}
```
Comment (vmg) [inline go/vt/schemadiff/mysql.go:94]: This is authoritatively available from https://github.com/vitessio/vitess/blob/main/go/mysql/collations/env.go#L73

**Review by vmg — CHANGES_REQUESTED** [review]
Looks good but I would use the authoritative list of collations from our `collations` package because it's going to be kept up to date more consistently. :)

File: go/vt/schemadiff/mysql.go (hunk 4 above)
Comment (dbussink) [reply go/vt/schemadiff/mysql.go]: Cleaned this up!

**Review by vmg — APPROVED** [review]
LGTM!

**Pull request merged**